File: src/main/java/xyz/chunkstories/core/entity/EntityPlayerController.java
package xyz.chunkstories.core.entity;

import org.joml.Vector3d;
import org.joml.Vector3dc;
import xyz.chunkstories.api.Location;
import xyz.chunkstories.api.client.LocalPlayer;
import xyz.chunkstories.api.entity.Controller;
import xyz.chunkstories.api.entity.Entity;
import xyz.chunkstories.api.entity.traits.TraitInteractible;
import xyz.chunkstories.api.entity.traits.serializable.TraitControllable;
import xyz.chunkstories.api.entity.traits.serializable.TraitCreativeMode;
import xyz.chunkstories.api.entity.traits.serializable.TraitSelectedItem;
import xyz.chunkstories.api.events.player.voxel.PlayerVoxelModificationEvent;
import xyz.chunkstories.api.exceptions.world.WorldException;
import xyz.chunkstories.api.input.Input;
import xyz.chunkstories.api.item.ItemVoxel;
import xyz.chunkstories.api.item.interfaces.ItemZoom;
import xyz.chunkstories.api.item.inventory.ItemPile;
import xyz.chunkstories.api.player.Player;
import xyz.chunkstories.api.sound.SoundSource;
import xyz.chunkstories.api.world.World;
import xyz.chunkstories.api.world.WorldClient;
import xyz.chunkstories.api.world.WorldMaster;
import xyz.chunkstories.api.world.cell.CellData;
import xyz.chunkstories.api.world.cell.FutureCell;
import xyz.chunkstories.core.CoreOptions;
import xyz.chunkstories.core.item.MiningProgress;
import xyz.chunkstories.core.item.inventory.CreativeInventoryHelperKt;

import java.util.Iterator;

class EntityPlayerController extends TraitControllable {

	private final EntityPlayer entityPlayer;

	public EntityPlayerController(EntityPlayer entity) {
		super(entity);
		entityPlayer = entity;
	}

	@Override public boolean onEachFrame() {
		Controller controller = getController();
		if (controller instanceof LocalPlayer && ((LocalPlayer) controller).hasFocus()) {
			moveCamera((LocalPlayer) controller);
			return true;
		}
		return false;
	}

	double lastPX = -1f;
	double lastPY = -1f;

	public void moveCamera(LocalPlayer controller) {
		if (entityPlayer.entityHealth.isDead())
			return;
		if (!controller.getInputsManager().getMouse().isGrabbed())
			return;

		double cPX = controller.getInputsManager().getMouse().getCursorX();
		double cPY = controller.getInputsManager().getMouse().getCursorY();

		double dx = 0, dy = 0;
		if (lastPX != -1f) {
			dx = cPX - controller.getWindow().getWidth() / 2.0;
			dy = cPY - controller.getWindow().getHeight() / 2.0;
		}
		lastPX = cPX;
		lastPY = cPY;

		double rotH = entityPlayer.entityRotation.getHorizontalRotation();
		double rotV = entityPlayer.entityRotation.getVerticalRotation();

		double modifier = 1.0f;
		ItemPile selectedItem = entityPlayer.traits.tryWith(TraitSelectedItem.class, eci -> eci.getSelectedItem());

		if (selectedItem != null && selectedItem.getItem() instanceof ItemZoom) {
			ItemZoom item = (ItemZoom) selectedItem.getItem();
			modifier = 1.0 / item.getZoomFactor();
		}

		rotH -= dx * modifier / 3f * controller.getClient().getConfiguration().getDoubleValue(CoreOptions.INSTANCE.getMouseSensitivity());
		rotV += dy * modifier / 3f * controller.getClient().getConfiguration().getDoubleValue(CoreOptions.INSTANCE.getMouseSensitivity());
		entityPlayer.entityRotation.setRotation(rotH, rotV);

		controller.getInputsManager().getMouse().setMouseCursorLocation(controller.getWindow().getWidth() / 2.0, controller.getWindow().getHeight() / 2.0);
	}

	/*@Override
	public void setupCamera(RenderingInterface renderer) {
		lastCameraLocation = getLocation();

		renderer.getCamera().setCameraPosition(new Vector3d(getLocation().add(0.0, stance.get().eyeLevel, 0.0)));

		renderer.getCamera().setRotationX(entityRotation.getVerticalRotation());
		renderer.getCamera().setRotationY(entityRotation.getHorizontalRotation());

		float modifier = 1.0f;
		if (selectedItemComponent.getSelectedItem() != null
				&& selectedItemComponent.getSelectedItem().getItem() instanceof ItemZoom) {
			ItemZoom item = (ItemZoom) selectedItemComponent.getSelectedItem().getItem();
			modifier = 1.0f / item.getZoomFactor();
		}

		float videoFov = (float) renderer.getClient().getConfiguration().getDoubleOption("client.video.fov");
		float speedEffect = (float) (entityVelocity.getVelocity().x() * entityVelocity.getVelocity().x()
				+ entityVelocity.getVelocity().z() * entityVelocity.getVelocity().z());

		speedEffect -= 0.07f * 0.07f;
		speedEffect = Math.max(0.0f, speedEffect);
		speedEffect *= 500.0f;

		renderer.getCamera().setFOV(modifier * (float) (videoFov + speedEffect));
	}*/

	@Override public boolean onControllerInput(Input input) {
		Controller controller = getController();

		// We are moving inventory bringup here !
		if (input.getName().equals("inventory") && entityPlayer.getWorld() instanceof WorldClient) {

			if (entityPlayer.creativeMode.get()) {
				((WorldClient) entityPlayer.getWorld()).getClient().getGui().openInventories(entityPlayer.inventory.getInventory(),
						CreativeInventoryHelperKt.createCreativeInventory(getEntity().getWorld().getContent().voxels()));
			} else {
				((WorldClient) entityPlayer.getWorld()).getClient().getGui()
						.openInventories(entityPlayer.inventory.getInventory(), entityPlayer.armor.getInventory());
			}

			return true;
		}

		Location blockLocation = entityPlayer.raytracer.getBlockLookingAt(true, false);

		double maxLen = 1024;

		if (blockLocation != null) {
			Vector3d diff = new Vector3d(blockLocation).sub(entityPlayer.getLocation());
			// Vector3d dir = diff.clone().normalize();
			maxLen = diff.length();
		}

		Vector3d initialPosition = new Vector3d(entityPlayer.getLocation());
		initialPosition.add(new Vector3d(0, entityPlayer.stance.getStance().getEyeLevel(), 0));

		Vector3dc direction = entityPlayer.entityRotation.getDirectionLookingAt();

		Iterator<Entity> i = entityPlayer.getWorld().getCollisionsManager().rayTraceEntities(initialPosition, direction, maxLen);
		while (i.hasNext()) {
			Entity e = i.next();
			if (e != entityPlayer && e.traits.with(TraitInteractible.class, ti -> ti.handleInteraction(entityPlayer, input)))
				return true;
		}

		ItemPile itemSelected = entityPlayer.selectedItemComponent.getSelectedItem();
		if (itemSelected != null) {
			// See if the item handles the interaction
			if (itemSelected.getItem().onControllerInput(entityPlayer, itemSelected, input, controller))
				return true;
		}
		if (entityPlayer.getWorld() instanceof WorldMaster) {
			// Creative mode features building and picking.
			if (entityPlayer.creativeMode.get()) {
				if (input.getName().equals("mouse.left")) {
					if (blockLocation != null) {
						// Player events mod
						if (controller instanceof Player) {
							Player player = (Player) controller;
							World.WorldCell cell = entityPlayer.getWorld().peekSafely(blockLocation);
							FutureCell future = new FutureCell(cell);
							future.setVoxel(entityPlayer.getDefinition().store().parent().voxels().air());
							future.setBlocklight(0);
							future.setSunlight(0);
							future.setMetaData(0);

							PlayerVoxelModificationEvent event = new PlayerVoxelModificationEvent(cell, future, TraitCreativeMode.Companion.getCREATIVE_MODE(),
									player);

							// Anyone has objections ?
							entityPlayer.getWorld().getGameContext().getPluginManager().fireEvent(event);

							if (event.isCancelled())
								return true;

							MiningProgress.Companion.spawnBlockDestructionParticles(blockLocation, entityPlayer.getWorld());
							entityPlayer.getWorld().getSoundManager()
									.playSoundEffect("sounds/gameplay/voxel_remove.ogg", SoundSource.Mode.NORMAL, blockLocation, 1.0f, 1.0f);

							try {
								entityPlayer.getWorld().poke(future, entityPlayer);
								// world.poke((int)blockLocation.x, (int)blockLocation.y, (int)blockLocation.z,
								// null, 0, 0, 0, this);
							} catch (WorldException e) {
								// Discard but maybe play some effect ?
							}

							return true;
						}
					}
				} else if (input.getName().equals("mouse.middle")) {
					if (blockLocation != null) {
						CellData ctx = entityPlayer.getWorld().peekSafely(blockLocation);

						if (!ctx.getVoxel().isAir()) {
							// Spawn new itemPile in his inventory
							ItemVoxel item = entityPlayer.getWorld().getGameContext().getContent().items().getItemDefinition("item_voxel").newItem();
							item.setVoxel(ctx.getVoxel());
							item.setVoxelMeta(ctx.getMetaData());

							//ItemPile itemVoxel = new ItemPile(item);
							//entityPlayer.inventory.setItemPileAt(entityPlayer.selectedItemComponent.getSelectedSlot(), 0, itemVoxel);
							entityPlayer.inventory.getInventory().setItemAt(entityPlayer.selectedItemComponent.getSelectedSlot(), 0, item);
							return true;
						}
					}
				}
			}
		}
		// Here goes generic entity response to interaction

		// n/a

		// Then we check if the world minds being interacted with
		return entityPlayer.getWorld().handleInteraction(entityPlayer, blockLocation, input);
	}
}

File: src/main/java/xyz/chunkstories/core/entity/EntityPlayerController.kt
package xyz.chunkstories.core.entity

import org.joml.Math
import org.joml.Vector3d
import xyz.chunkstories.api.client.IngameClient
import xyz.chunkstories.api.client.LocalPlayer
import xyz.chunkstories.api.entity.traits.TraitInteractible
import xyz.chunkstories.api.entity.traits.serializable.*
import xyz.chunkstories.api.graphics.structs.Camera
import xyz.chunkstories.api.graphics.structs.makeCamera
import xyz.chunkstories.api.input.Input
import xyz.chunkstories.api.item.ItemVoxel
import xyz.chunkstories.api.item.interfaces.ItemZoom
import xyz.chunkstories.api.physics.RayResult
import xyz.chunkstories.api.util.kotlin.toVec3f
import xyz.chunkstories.api.world.WorldClient
import xyz.chunkstories.api.world.WorldMaster
import xyz.chunkstories.api.world.chunk.ChunkCell
import xyz.chunkstories.core.CoreOptions
import xyz.chunkstories.core.gui.CreativeBlockSelector
import java.lang.Exception

internal class EntityPlayerController(private val entityPlayer: EntityPlayer) : TraitControllable(entityPlayer) {

    var lastPX = -1.0
    var lastPY = -1.0

    override fun onEachFrame(): Boolean {
        val controller = controller
        if (controller is LocalPlayer && controller.hasFocus()) {
            rotateCameraAccordingToMouse(controller)
            return true
        }
        return false
    }

    fun rotateCameraAccordingToMouse(controller: LocalPlayer) {
        if (entityPlayer.traitHealth.isDead)
            return
        if (!controller.inputsManager.mouse.isGrabbed)
            return

        val cPX = controller.inputsManager.mouse.cursorX
        val cPY = controller.inputsManager.mouse.cursorY

        var dx = 0.0
        var dy = 0.0
        if (lastPX != -1.0) {
            dx = cPX - controller.window.width / 2.0
            dy = cPY - controller.window.height / 2.0
        }
        lastPX = cPX
        lastPY = cPY

        var rotH = entityPlayer.traitRotation.horizontalRotation.toDouble()
        var rotV = entityPlayer.traitRotation.verticalRotation.toDouble()

        var modifier = 1.0
        val selectedItem = entityPlayer.traits[TraitSelectedItem::class]?.selectedItem

        if (selectedItem != null && selectedItem.item is ItemZoom) {
            val item = selectedItem.item as ItemZoom
            modifier = 1.0 / item.zoomFactor
        }

        rotH -= dx * modifier / 3f * controller.client.configuration.getDoubleValue(CoreOptions.mouseSensitivity)
        rotV += dy * modifier / 3f * controller.client.configuration.getDoubleValue(CoreOptions.mouseSensitivity)
        entityPlayer.traitRotation.setRotation(rotH, rotV)

        controller.inputsManager.mouse.setMouseCursorLocation(controller.window.width / 2.0, controller.window.height / 2.0)
    }

    override val camera: Camera
        get() {
            val client = entity.world.gameContext as? IngameClient ?: throw Exception("calling getCamera() on a non-client context is undefined behavior")

            val location = entity.location
            val cameraPosition = Vector3d(location)
            cameraPosition.y += entityPlayer.traitStance.stance.eyeLevel

            val direction = (entity.traits[TraitRotation::class]?.directionLookingAt ?: Vector3d(0.0, 0.0, 1.0)).toVec3f()
            val up = (entity.traits[TraitRotation::class]?.upDirection ?: Vector3d(0.0, 0.0, 1.0)).toVec3f()

            val fovModifier = entityPlayer.traitSelectedItem.selectedItem?.let { (it.item as? ItemZoom)?.zoomFactor } ?: 1f
            var speedEffect = (entityPlayer.traitVelocity.velocity.x() * entityPlayer.traitVelocity.velocity.x() + entityPlayer.traitVelocity.velocity.z() * entityPlayer.traitVelocity.velocity.z()).toFloat()
            speedEffect -= 0.07f * 0.07f
            speedEffect = Math.max(0.0f, speedEffect)
            speedEffect *= 50.0f

            return client.makeCamera(cameraPosition, direction, up, fovModifier * (90f + speedEffect))
        }

    override fun onControllerInput(input: Input): Boolean {
        val controller = controller

        // We are moving inventory bringup here !
        if (input.name == "inventory" && entityPlayer.world is WorldClient) {

            if (entityPlayer.traitCreativeMode.get()) {
                entityPlayer.world.client.gui.let {
                    it.topLayer = CreativeBlockSelector(it, it.topLayer)
                }
                //entityPlayer.world.client.gui.openInventories(entityPlayer.traits[TraitInventory::class]?.inventory!!, entity.world.content.voxels().createCreativeInventory())
            } else {
                entityPlayer.world.client.gui.openInventories(entityPlayer.traits[TraitInventory::class]?.inventory!!)
            }

            return true
        }


        //var maxLen = 1024.0
        //val diff = Vector3d(lookingAt).sub(entityPlayer.location)
        //maxLen = diff.length()

        /*val initialPosition = Vector3d(entityPlayer.location)
        initialPosition.add(Vector3d(0.0, entityPlayer.traitStance.stance.eyeLevel, 0.0))

        val direction = entityPlayer.traitRotation.directionLookingAt*/

        /*val ray = RayQuery(Location(entityPlayer.world, initialPosition), direction, 0.0, 256.0)
        val pointingAt = ray.trace()
        println(pointingAt)*/

        /*val i = entityPlayer.world.collisionsManager.rayTraceEntities(initialPosition, direction, maxLen)
        while (i.hasNext()) {
            val e = i.next()
            if (e !== entityPlayer && e.traits[TraitInteractible::class]?.handleInteraction(entityPlayer, input) == true)
                return true
        }*/// Spawn new itemPile in his inventory
        //val item = entityPlayer.world.gameContext.content.items().getItemDefinition("item_voxel")!!.newItem<ItemVoxel>()
        //item.voxel = voxel
        //item.voxelMeta = peekedCell.metaData
        // Here goes generic entity response to interaction

        // n/a

        // Then we check if the world minds being interacted with
        // Creative mode features building and picking.

        val lookingAt = entityPlayer.traitSight.getLookingAt(5.0)
        when (lookingAt) {
            is RayResult.Hit.EntityHit -> {
                val observedEntity = lookingAt.entity
                if(observedEntity != entity && observedEntity.traits[TraitInteractible::class]?.handleInteraction(entity, input) == true) {
                    return true
                }
            }
            is RayResult.Hit.VoxelHit -> {
                val cell = lookingAt.cell
                // Should always be true !
                if(cell is ChunkCell) {
                    if(cell.voxel.handleInteraction(entity, cell, input)) {
                        return true
                    }
                }
            }
        }

        val itemInHand = entityPlayer.traitSelectedItem.selectedItem
        if (itemInHand != null) {
            // See if the item handles the interaction
            if (itemInHand.item.onControllerInput(entityPlayer, itemInHand, input, controller!!))
                return true
        } else {

        }

        if (entityPlayer.world is WorldMaster && lookingAt is RayResult.Hit.VoxelHit) {
            // Creative mode features building and picking.
            if (entityPlayer.traitCreativeMode.get()) {
                if (input.name == "mouse.left") {
                    val cell = lookingAt.cell
                    cell.voxel.breakBlock(cell, TraitCreativeMode.CREATIVE_MODE_MINING_TOOL, entity)
                } else if (input.name == "mouse.middle") {
                    val peekedCell = lookingAt.cell
                    val voxel = peekedCell.voxel

                    if (!voxel.isAir()) {
                        // Spawn new itemPile in his inventory
                        val item = voxel.getVariant(peekedCell).newItem<ItemVoxel>()
                        entityPlayer.traits[TraitInventory::class]?.inventory!!.setItemAt(entityPlayer.traitSelectedItem.getSelectedSlot(), 0, item, 1, force = true)
                        return true
                    }
                }
            }
        }

        return false
    }
}
